### PR TITLE
Support continuous batching in sequence batch streaming case

### PR DESCRIPTION
**Pull request by lxning**
## Description

Please read our [CONTRIBUTING.md](https://github.com/pytorch/serve/blob/master/CONTRIBUTING.md) prior to creating your first pull request.

Please include a summary of the feature or issue being fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.

Fixes #(issue)

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] This change requires a documentation update

## Feature/Issue validation/testing

Please describe the Unit or Integration tests that you ran to verify your changes and relevant result summary. Provide instructions so it can be reproduced.
Please also list any relevant details for your test configuration.

```
 pytest test_example_stateful_sequence_continuous_batching_http.py
============================== test session starts ===============================
platform linux -- Python 3.10.14, pytest-7.3.1, pluggy-1.4.0
rootdir: /home/ubuntu/serve
plugins: cov-4.1.0, mock-3.12.0
collected 3 items

test_example_stateful_sequence_continuous_batching_http.py ...             [100%]

=============================== 3 passed in 19.81s ===============================
```

## Checklist:

- [ ] Did you have fun?
- [ ] Have you added tests that prove your fix is effective or that this feature works?
- [ ] Has code been commented, particularly in hard-to-understand areas?
- [ ] Have you made corresponding changes to the documentation?

### CODE REVIEW

File: test/pytest/test_example_stateful_sequence_continuous_batching_http.py [hunk 1]
```diff
@@ -0,0 +1,363 @@
...
+            elif sequence_id == "seq_1":
+                headers_seq_1 = {
+                    "ts_request_sequence_id": s_id,
+                }
+            start = False
+        else:
+            response = requests.post(
+                url=f"http://localhost:8080/predictions/{model_name}",
+                headers=headers_seq_0 if sequence_id == "seq_0" else headers_seq_1,
+                data=str(idx + 1).encode(),
```
Comment (mreso) [inline test/pytest/test_example_stateful_sequence_continuous_batching_http.py:312]: For seq_0 we expect the last value to stay constant as we send 0, but we also send 0 to seq_1. Why do we expecty the last value to be 30 there?

File: test/pytest/test_example_stateful_sequence_continuous_batching_http.py [hunk 2]
```diff
@@ -0,0 +1,363 @@
...
+    for idx in range(5):
+        if idx == 2:
+            cancel = True
+        if sequence_id == "seq_0":
+            idx = 2 * idx
+        elif sequence_id == "seq_1":
+            idx = 2 * idx + 1
+
+        if cancel is True and sequence_id == "seq_0":
+            response = requests.post(
```
Comment (mreso) [inline test/pytest/test_example_stateful_sequence_continuous_batching_http.py:333]: This test does not cover the continuous batching part. We should add that in the test.

**Review by mreso — CHANGES_REQUESTED** [review]
Left some comments. Please add continuous batching to the test to make sure this part works as well.

File: examples/stateful/sequence_continuous_batching/stateful_handler.py [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+        """
+
+        return data
+
+    def clean_up_seq(self, seq_id):
+        # clean up
+        del self.cache[seq_id]
+        del self.context.cache[seq_id]
+
+    def clean_up_req(self, seq_id, req_id):
```
[the remark this answers is not in the document]
Comment (lxning) [reply examples/stateful/sequence_continuous_batching/stateful_handler.py:129]: they are used to clean up cache in 2 cases:
1. session end: clean this session in context cache
2. cancel: clean req in context cache for this session

File: frontend/server/src/main/java/org/pytorch/serve/util/messages/RequestInput.java [hunk 4]
```diff
@@ -75,10 +79,10 @@ public void setClientExpireTS(long clientTimeoutInMills) {
     }
 
     public String getSequenceId() {
-        if (sequenceId == null) {
+        if (sequenceId.isEmpty()) {
```
[the remark this answers is not in the document]
Comment (lxning) [reply frontend/server/src/main/java/org/pytorch/serve/util/messages/RequestInput.java:82]: I keep the code of setting requestinput as the original in http handler part. this function is only called by sequence batching case. If the sequence is already set as non-empty value, we don't need to check hash table

File: frontend/server/src/main/java/org/pytorch/serve/wlm/SequenceContinuousBatching.java [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+                    jobGroup.setFinished(true);
+                }
+            }
+        }
+    }
+
+    @Override
+    protected void pollInferJob() throws InterruptedException {
+        // TBD: Temporarily hard code the continuous batch size is 2 * batchSize
+        model.pollInferJob(jobs, model.getBatchSize() * 2 - jobs.size(), jobsQueue);
```
[the remark this answers is not in the document]
Comment (lxning) [reply frontend/server/src/main/java/org/pytorch/serve/wlm/SequenceContinuousBatching.java:122]: No, it is not necessary for backend handler to sort b/c backend handler preprocessing is able to mark the previous running request as "cancel"

File: test/pytest/test_example_stateful_sequence_continuous_batching_http.py (hunk 1 above)
Comment (lxning) [reply test/pytest/test_example_stateful_sequence_continuous_batching_http.py:312]: b/c seq_1 call infer_stateful, not infer_stateful_end

File: test/pytest/test_example_stateful_sequence_continuous_batching_http.py (hunk 2 above)
Comment (lxning) [reply test/pytest/test_example_stateful_sequence_continuous_batching_http.py:333]: it is complicate to match the final result b/c it is dynamically in continuous batching case. Here, the client side does not enable chunked encode, which mean skipping the rest of response, to get the fixed the result.

File: frontend/server/src/main/java/org/pytorch/serve/wlm/SequenceContinuousBatching.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (mreso) [reply frontend/server/src/main/java/org/pytorch/serve/wlm/SequenceContinuousBatching.java:122]: @lxning So you mean the preprocessing in the handler will need to sort this out as there can be multiple requests from multiple sequences and in no specific pattern. In the end the preprocessing in the handler will need to go though the batch and see if besides the previous request another cancel request is in the batch as well and than cancel it/clean it up.

File: test/pytest/test_example_stateful_sequence_continuous_batching_http.py (hunk 2 above)
Comment (mreso) [reply test/pytest/test_example_stateful_sequence_continuous_batching_http.py:333]: You can just use sleep in the inference() call to make sure we get a cancellation request in after we get a certain response through the chunked response.

File: test/pytest/test_example_stateful_sequence_continuous_batching_http.py (hunk 1 above)
Comment (mreso) [reply test/pytest/test_example_stateful_sequence_continuous_batching_http.py:312]: got it, missed that, thanks

File: frontend/server/src/main/java/org/pytorch/serve/wlm/WorkLoadManager.java [hunk 6]
```diff
@@ -229,12 +229,10 @@ private void addThreads(
 
             BatchAggregator aggregator;
 
-            if (model.isSequenceBatching()) {
-                if (model.isSequenceContinuousBatch()) {
-                    aggregator = new SequenceContinuousBatching(model);
-                } else {
-                    aggregator = new SequenceBatching(model);
-                }
+            if (model.isSequenceContinuousBatch()) {
```
Comment (mreso) [inline frontend/server/src/main/java/org/pytorch/serve/wlm/WorkLoadManager.java:232]: This is incorrect as it tests for isSequenceContinuousBatch in both cases. I also meant that we should not need to create the separate configuration keyword sequenceContinuousBatching but we can just rely on sequenceBatching + continuousBatching being set to True.

**Review by mreso — CHANGES_REQUESTED** [review]
Please see comments.

File: frontend/server/src/main/java/org/pytorch/serve/wlm/SequenceContinuousBatching.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (lxning) [reply frontend/server/src/main/java/org/pytorch/serve/wlm/SequenceContinuousBatching.java:163]: No, there is no sort. it is based on "[num_requests](https://github.com/pytorch/serve/blob/4b43233381beec5034f4c2c65dc162d11e833fa5/examples/stateful/sequence_continuous_batching/stateful_handler.py#L185)"

File: frontend/server/src/main/java/org/pytorch/serve/wlm/WorkLoadManager.java (hunk 6 above)
Comment (lxning) [reply frontend/server/src/main/java/org/pytorch/serve/wlm/WorkLoadManager.java]: remove model config "sequenceContinuousBatching"

File: examples/stateful/sequence_continuous_batching/stateful_handler.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (lxning) [reply examples/stateful/sequence_continuous_batching/stateful_handler.py]: combine the 2 function as one.

**Review by mreso — APPROVED** [review]
LGTM

**Pull request merged**